src/app/dataset-view/additional-components/metadata-component/components/set-transform/components/search-section/search-section.component.ts

```diff
@@ -0,0 +1,112 @@
+import { SearchApi } from "src/app/api/search.api";
+import { MaybeNull } from "src/app/common/app.types";
+import AppValues from "src/app/common/app.values";
+import { DatasetService } from "src/app/dataset-view/dataset.service";
+import { DatasetSchema } from "src/app/interface/dataset.interface";
+import { DatasetAutocompleteItem } from "src/app/interface/search.interface";
+import { DatasetNode } from "../../set-transform.types";
+import { BaseComponent } from "src/app/common/base.component";
+
+@Component({
+    selector: "app-search-section",
+    templateUrl: "./search-section.component.html",
+    styleUrls: ["./search-section.component.sass"],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SearchSectionComponent extends BaseComponent {
+    public searchDataset = "";
+    private readonly delayTime: number = AppValues.SHORT_DELAY_MS;
+    @Input() public inputDatasets: Set<string>;
+
+    public treeControl = new NestedTreeControl<DatasetNode>(
+        (node) => node.children,
+    );
+    @Input() public dataSource: MatTreeNestedDataSource<DatasetNode>;
+    @Input() public TREE_DATA: DatasetNode[];
+    constructor(
+        private appSearchAPI: SearchApi,
+        private datasetService: DatasetService,
+    ) {
+        super();
+    }
+
+    public search: OperatorFunction<
+        string,
+        readonly DatasetAutocompleteItem[]
+    > = (text$: Observable<string>) => {
+        return text$.pipe(
+            debounceTime(this.delayTime),
+            distinctUntilChanged(),
+            switchMap((term: string) =>
+                this.appSearchAPI.autocompleteDatasetSearch(term),
+            ),
+        );
+    };
+
+    public formatter(x: DatasetAutocompleteItem | string): string {
+        return typeof x !== "string" ? (x.dataset.name as string) : x;
+    }
+
+    public onSelectItem(event: NgbTypeaheadSelectItemEvent): void {
+        const value = event.item as DatasetAutocompleteItem;
+        const id = value.dataset.id as string;
+        const name = value.dataset.name as string;
+        const inputDataset = JSON.stringify({
+            id,
+            name,
+        });
+        if (
+            value.__typename !== "all" &&
+            !this.inputDatasets.has(inputDataset)
+        ) {
+            this.inputDatasets.add(inputDataset);
+            this.trackSubscription(
+                this.datasetService
+                    .requestDatasetSchema(id)
+                    .subscribe((data: GetDatasetSchemaQuery) => {
+                        if (data.datasets.byId) {
+                            const schema: MaybeNull<DatasetSchema> = data
+                                .datasets.byId.metadata.currentSchema
+                                ? (JSON.parse(
+                                      data.datasets.byId.metadata.currentSchema
+                                          .content,
+                                  ) as DatasetSchema)
```

Comment:
I think we have similar code already in other parts of the system, so please extract a common utility helper.

Reply:
Move this code to helper file